### PR TITLE
fix angle errors

### PR DESCRIPTION
**Pull request by Ericwang6**
Hi Rohith,

I think I fixed most of the bugs you mentioned (and I found) in your angle cuda code. I didn't modify the CPU code, so the CPU tests still cannot pass now.

### CODE REVIEW

File: csrc/angle/harmonic_angle_cuda.cu [hunk 1]
```diff
@@ -79,29 +79,25 @@ __global__ void harmonic_angle_energy_grad_cuda_kernel(
...
-    scalar_t g2x = prefix * (-v1x / v1_norm - v2x / v2_norm + cos_theta * (v1x / v1_norm + v2x / v2_norm));
-    scalar_t g2y = prefix * (-v1y / v1_norm - v2y / v2_norm + cos_theta * (v1y / v1_norm + v2y / v2_norm));
-    scalar_t g2z = prefix * (-v1z / v1_norm - v2z / v2_norm + cos_theta * (v1z / v1_norm + v2z / v2_norm));
+    scalar_t g3x = prefix * (v1x - cos_theta * v2x / v2_norm * v1_norm);
+    scalar_t g3y = prefix * (v1y - cos_theta * v2y / v2_norm * v1_norm);
+    scalar_t g3z = prefix * (v1z - cos_theta * v2z / v2_norm * v1_norm);
 
-    scalar_t g3x = prefix * (v1x / v2_norm - cos_theta * v2x / v2_norm);
-    scalar_t g3y = prefix * (v1y / v2_norm - cos_theta * v2y / v2_norm);
-    scalar_t g3z = prefix * (v1z / v2_norm - cos_theta * v2z / v2_norm);
```
Comment (Ericwang6) [inline csrc/angle/harmonic_angle_cuda.cu:92]: I think here you compute the gradients incorrectly. Here is a derivation following your notation.

$$\cos\theta=\frac{v_1\cdot v_2}{|v_1| |v_2|}$$

$$\frac{\partial\cos\theta}{\partial v_1}=\frac{1}{|v_2|}\frac{v_2\cdot|v_1|-(v_1\cdot v_2)v_1/|v_1|}{|v_1|^2}=\frac{1}{|v_1||v_2|}\left(v_2-\frac{v_1\cdot v_2}{|v_1|^2}v_1\right)=\frac{1}{|v_1||v_2|}\left(v_2-\frac{\cos\theta |v_2|}{|v_1|}v_1\right)$$

File: csrc/angle/harmonic_angle_cuda.cu [hunk 2]
```diff
@@ -79,29 +79,25 @@ __global__ void harmonic_angle_energy_grad_cuda_kernel(
...
+    atomicAdd(&coord_grad[offset_0],     g1x);
+    atomicAdd(&coord_grad[offset_0 + 1], g1y);
+    atomicAdd(&coord_grad[offset_0 + 2], g1z);
```
Comment (Ericwang6) [inline csrc/angle/harmonic_angle_cuda.cu:92]: Since we are calculating gradients, not forces ($F=-\partial E/\partial r$), we don't need a minus sign before `g1x`

File: csrc/angle/harmonic_angle_cuda.cu [hunk 3]
```diff
@@ -79,29 +79,25 @@ __global__ void harmonic_angle_energy_grad_cuda_kernel(
...
 
-    atomicAdd(&coord_grad[offset_0],     -g1x);
-    atomicAdd(&coord_grad[offset_0 + 1], -g1y);
-    atomicAdd(&coord_grad[offset_0 + 2], -g1z);
+    atomicAdd(&coord_grad[offset_1],     -g1x-g3x);
+    atomicAdd(&coord_grad[offset_1 + 1], -g1y-g3y);
+    atomicAdd(&coord_grad[offset_1 + 2], -g1z-g3z);
```
Comment (Ericwang6) [inline csrc/angle/harmonic_angle_cuda.cu:96]: There is a easy way to calculate gradients for the central atom. It's just `-g1-g3` because of the energy is conservative.

File: csrc/angle/harmonic_angle_cuda.cu [hunk 4]
```diff
@@ -146,7 +142,7 @@ std::tuple<at::Tensor, at::Tensor, at::Tensor> compute_harmonic_angle_energy_gra
 ) {
     int64_t nangles = triplets.size(0);
 
-    int block_dim = 1024;
+    int block_dim = 512;
```
Comment (Ericwang6) [inline csrc/angle/harmonic_angle_cuda.cu:145]: Somehow I found the cuda cannot maintain `blockDim = 1024` when double precision is used. I assume it's because there will not have that enough register allocated for each thread. `blockDim = 512` will work. There are be some ways to rewrite the cuda kernel to not using too many registers, but we can leave it now I think.

File: csrc/angle/harmonic_angle_interface.cpp [hunk 5]
```diff
@@ -1,7 +1,7 @@
 #include <pybind11/pybind11.h>
 #include <torch/library.h>
 
-TORCH_LIBRARY(torchff, m) {
+TORCH_LIBRARY_FRAGMENT(torchff, m) {
```
Comment (Ericwang6) [inline csrc/angle/harmonic_angle_interface.cpp:4]: I asked GPT and it seems that here is the reason to cause the error. We already have a `TORCH_LIBRARY` in harmonic bond under namespace `torchff`, so here it needs to be changed into `TORCH_LIBRARY_FRAGMENT` .

GPT also suggested another way to fix this, which is to put all the `TORCH_LIBRARY` macro in one `interface.cpp` file. But I thought that will change the code structure too much and make it less modular.

File: tests/test_angle.py [hunk 6]
```diff
@@ -5,7 +5,7 @@
 import torchff
 
 def harmonic_angle(coords, angles, theta0, k):
-    v1 = coords[angles[:, 1]] - coords[angles[:, 0]]
+    v1 = coords[angles[:, 0]] - coords[angles[:, 1]]
```
Comment (Ericwang6) [inline tests/test_angle.py:8]: A typo here.

File: tests/test_angle.py [hunk 7]
```diff
@@ -17,8 +17,8 @@ def harmonic_angle(coords, angles, theta0, k):
     return torch.sum(ene)
 
 @pytest.mark.parametrize("device, dtype", [
-    ('cpu', torch.float64), 
-    ('cpu', torch.float32), 
+    # ('cpu', torch.float64), 
+    # ('cpu', torch.float32),
```
Comment (Ericwang6) [inline tests/test_angle.py:21]: Please fix the CPU code. I did not change your CPU code.

File: tests/test_angle.py [hunk 8]
```diff
@@ -46,8 +46,9 @@ def test_harmonic_angle(device, dtype):
     ene.backward()
     grads = [coords.grad.clone().detach()]
 
-    for c, g, gref in zip(['coord'], grads, grads_ref):
-        assert torch.allclose(g, gref, atol=1e-5), f'Gradient {c} not the same'
+    grad_atol = 1e-5 if dtype is torch.float64 else 1e-2
+    for c, g, gref in zip(['coord', 'theta0', 'k'], grads, grads_ref):
+        assert torch.allclose(g, gref, atol=grad_atol), f'Gradient {c} not the same'
```
Comment (Ericwang6) [inline tests/test_angle.py:51]: Here I specified a tolerance in `torch.allclose` for different precisions. It seems that the single precision needs a really loose criteria.

File: torchff/angle.py [hunk 9]
```diff
@@ -14,7 +14,7 @@ def forward(ctx, coords, triplets, theta0, k):
     def backward(ctx, grad_outputs):
         coords, triplets, theta0, k = ctx.saved_tensors
         grad_outputs = torch.ops.torchff.compute_harmonic_angle_energy_grad(coords, triplets, theta0, k)
-        return grad_coords[0], None, grad_outputs[1], grad_outputs[2]
+        return grad_outputs[0], None, grad_outputs[1], grad_outputs[2]
```
Comment (Ericwang6) [inline torchff/angle.py:17]: A naming typo here.
